eth/plasma.go

```diff
@@ -0,0 +1,241 @@
+		CallOpts: bind.CallOpts{
+			Pending: true,
+		},
+		TransactOpts: bind.TransactOpts{
+			From:     auth.From,
+			Signer:   auth.Signer,
+			GasLimit: 3141592, // aribitrary. TODO: check this
+		},
+	}
+
+	plasma := &Plasma{
+		session: plasmaSession,
+		client:  client,
+		// capacity argument is advisory and not enforced in the memdb implementation
+		// TODO: flush the in-memory DB to a local one to bound memory consumption
+		memdb:  memdb.New(comparer.DefaultComparer, 1),
+		logger: logger,
+
+		ethBlockNum:   big.NewInt(0),
+		blockNum:      sdk.ZeroUint(),
+		finalityBound: finalityBound,
+
+		lock: &sync.Mutex{},
+	}
+
+	ethCh, err := plasma.client.SubscribeToHeads()
+	if err != nil {
+		plasma.logger.Error("Could not successfully subscribe to heads: %v", err)
+		return nil, err
+	}
+
+	// start listeners
+	go watchEthBlocks(plasma, ethCh)
+	go watchDeposits(plasma)
+	go watchExits(plasma)
+
+	return plasma, nil
+}
+
+// SubmitBlock proxy. TODO: handle batching with a timmer interrupt
+func (plasma *Plasma) SubmitBlock(header []byte, numTxns sdk.Uint, fee sdk.Uint) (*types.Transaction, error) {
+	plasma.blockNum = plasma.blockNum.Add(sdk.OneUint())
+
+	tx, err := plasma.session.SubmitBlock(
+		header,
+		[]*big.Int{numTxns.BigInt()},
+		[]*big.Int{fee.BigInt()},
+		plasma.blockNum.BigInt())
+
+	if err != nil {
+		return nil, err
+	}
+
+	return tx, nil
+}
+
+// GetDeposit checks the existence of a deposit nonce
+func (plasma *Plasma) GetDeposit(nonce sdk.Uint) (*plasmaTypes.Deposit, error) {
+	key := prefixKey(depositPrefix, nonce.BigInt().Bytes())
+	data, err := plasma.memdb.Get(key)
+
+	var deposit plasmaTypes.Deposit
+	// check against the contract if the deposit is not in the cache or decoding fail
+	if err != nil && json.Unmarshal(data, &deposit) != nil {
+		if plasma.memdb.Contains(key) {
+			plasma.logger.Info("corrupted deposit found within db")
+			plasma.memdb.Delete(key)
+		}
+
+		d, err := plasma.session.Deposits(nonce.BigInt())
+		if err != nil {
+			plasma.logger.Error("contract call, deposits, failed")
+			return nil, err
+		}
+
+		if d.CreatedAt.Sign() == 0 {
+			return nil, errors.New("deposit does not exist")
+		}
+
+		deposit = plasmaTypes.Deposit{
+			Owner:    d.Owner,
+			Amount:   sdk.NewUintFromBigInt(d.Amount),
+			BlockNum: sdk.NewUintFromBigInt(d.EthBlocknum),
+		}
+	}
+
+	// save to the db
+	data, err = json.Marshal(deposit)
+	if err != nil {
+		plasma.logger.Error("error encoding deposit. will not be cached")
+	} else {
+		plasma.memdb.Put(key, data)
+	}
+
+	// check finality bound for the deposit
+	plasma.lock.Lock()
+	ethBlockNum := plasma.ethBlockNum
+	plasma.lock.Unlock()
+	if new(big.Int).Sub(ethBlockNum, deposit.BlockNum.BigInt()).Int64() >= plasma.finalityBound {
+		return &deposit, nil
+	} else {
+		return nil, errors.New("deposit not finalized")
+	}
+}
+
+// HasTXBeenExited indicates if the position has ever been exited
+func (plasma *Plasma) HasTXBeenExited(position [4]sdk.Uint) bool {
```

Comment:
This function should have the same semantics as GetDeposit.

Looks like this function will only track Exits since the node has started up, whereas GetDeposit can retreive full deposit history.

Rather than just checking for existence in memdb, function should make call to `plasma.session.Exits` analogous to `plasma.session.Deposits` so that full exit history is available for node to see